<header><h1>Promise.prototype.then()</h1>
<details class="baseline-indicator high"><summary><div class="status-title">Baseline <span class="not-bold">Widely available</span>
</div>

</summary><div class="extra">
<p>This feature is well established and works across many devices and browser versions. It’s been available across browsers since July 2015.</p>
<ul>
<li><a href="https://developer.mozilla.org/en-US/docs/Glossary/Baseline/Compatibility" data-glean="baseline_link_learn_more" target="_blank" class="learn-more">Learn more</a></li>
<li><a href="#browser_compatibility" data-glean="baseline_link_bcd_table">See full compatibility</a></li>
<li><a href="https://survey.alchemer.com/s3/7634825/MDN-baseline-feedback?page=%2Fen-US%2Fdocs%2FWeb%2FJavaScript%2FReference%2FGlobal_Objects%2FPromise%2Fthen&amp;level=high" data-glean="baseline_link_feedback" class="feedback-link" target="_blank" rel="noreferrer">Report feedback</a></li>
</ul>
</div></details></header><div class="section-content"><p>The <code>then()</code> method of <a href="../promise.html"><code>Promise</code></a> instances takes up to two arguments: callback functions for the fulfilled and rejected cases of the <code>Promise</code>. It stores the callbacks within the promise it is called on and immediately returns another <a href="../promise.html"><code>Promise</code></a> object, allowing you to <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Using_promises#chaining">chain</a> calls to other promise methods.</p></div>
<section aria-labelledby="try_it"><h2 id="try_it">Try it</h2>
<div class="section-content"><iframe class="interactive is-js-height" height="200" src="https://interactive-examples.mdn.mozilla.net/pages/js/promise-then.html" title="MDN Web Docs Interactive Example" allow="clipboard-write" loading="lazy"></iframe></div></section><section aria-labelledby="syntax"><h2 id="syntax">Syntax</h2>
<div class="section-content"><div class="code-example"><pre data-language="js">then(onFulfilled)
then(onFulfilled, onRejected)
</pre></div></div></section><section aria-labelledby="parameters"><h3 id="parameters">Parameters</h3>
<div class="section-content"><dl> <dt id="onfulfilled"><a href="#onfulfilled"><code>onFulfilled</code></a></dt> <dd> <p>A function to asynchronously execute when this promise becomes fulfilled. Its return value becomes the fulfillment value of the promise returned by <code>then()</code>. The function is called with the following arguments:</p> <dl> <dt id="value"><a href="#value"><code>value</code></a></dt> <dd> <p>The value that the promise was fulfilled with.</p> </dd> </dl> <p>If it is not a function, it is internally replaced with an <em>identity</em> function (<code>(x) =&gt; x</code>) which simply passes the fulfillment value forward.</p> </dd> <dt id="onrejected"><a href="#onrejected"><code>onRejected</code> <span class="badge inline optional">Optional</span></a></dt> <dd> <p>A function to asynchronously execute when this promise becomes rejected. Its return value becomes the fulfillment value of the promise returned by <code>then()</code>. The function is called with the following arguments:</p> <dl> <dt id="reason"><a href="#reason"><code>reason</code></a></dt> <dd> <p>The value that the promise was rejected with.</p> </dd> </dl> <p>If it is not a function, it is internally replaced with a <em>thrower</em> function (<code>(x) =&gt; { throw x; }</code>) which throws the rejection reason it received.</p> </dd> </dl></div></section><section aria-labelledby="return_value"><h3 id="return_value">Return value</h3>
<div class="section-content">
<p>Returns a new <a href="../promise.html"><code>Promise</code></a> immediately. This new promise is always pending when returned, regardless of the current promise's status.</p> <p>One of the <code>onFulfilled</code> and <code>onRejected</code> handlers will be executed to handle the current promise's fulfillment or rejection. The call always happens asynchronously, even when the current promise is already settled. The behavior of the returned promise (call it <code>p</code>) depends on the handler's execution result, following a specific set of rules. If the handler function:</p> <ul> <li>returns a value: <code>p</code> gets fulfilled with the returned value as its value.</li> <li>doesn't return anything: <code>p</code> gets fulfilled with <code>undefined</code> as its value.</li> <li>throws an error: <code>p</code> gets rejected with the thrown error as its value.</li> <li>returns an already fulfilled promise: <code>p</code> gets fulfilled with that promise's value as its value.</li> <li>returns an already rejected promise: <code>p</code> gets rejected with that promise's value as its value.</li> <li>returns another pending promise: <code>p</code> is pending and becomes fulfilled/rejected with that promise's value as its value immediately after that promise becomes fulfilled/rejected.</li> </ul>
</div></section><section aria-labelledby="description"><h2 id="description">Description</h2>
<div class="section-content">
<p>The <code>then()</code> method schedules callback functions for the eventual completion of a Promise — either fulfillment or rejection. It is the primitive method of promises: the <a href="../promise.html#thenables">thenable</a> protocol expects all promise-like objects to expose a <code>then()</code> method, and the <a href="catch.html"><code>catch()</code></a> and <a href="finally.html"><code>finally()</code></a> methods both work by invoking the object's <code>then()</code> method.</p> <p>For more information about the <code>onRejected</code> handler, see the <a href="catch.html"><code>catch()</code></a> reference.</p> <p><code>then()</code> returns a new promise object but mutates the promise object it's called on, appending the handlers to an internal list. Therefore the handler is retained by the original promise and its lifetime is at least as long as the original promise's lifetime. For example, the following example will eventually run out of memory even though the returned promise is not retained:</p> <div class="code-example"><pre data-language="js">const pendingPromise = new Promise(() =&gt; {});
while (true) {
  pendingPromise.then(doSomething);
}
</pre></div> <p>If you call the <code>then()</code> method twice on the same promise object (instead of chaining), then this promise object will have two pairs of settlement handlers. All handlers attached to the same promise object are always called in the order they were added. Moreover, the two promises returned by each call of <code>then()</code> start separate chains and do not wait for each other's settlement.</p> <p><a href="../promise.html#thenables">Thenable</a> objects that arise along the <code>then()</code> chain are always <a href="promise.html#the_resolve_function">resolved</a> — the <code>onFulfilled</code> handler never receives a thenable object, and any thenable returned by either handler are always resolved before being passed to the next handler. This is because when constructing the new promise, the <code>resolve</code> and <code>reject</code> functions passed by the <code>executor</code> are saved, and when the current promise settles, the respective function will be called with the fulfillment value or rejection reason. The resolving logic comes from the <code>resolve</code> function passed by the <a href="promise.html"><code>Promise()</code></a> constructor.</p> <p><code>then()</code> supports subclassing, which means it can be called on instances of subclasses of <code>Promise</code>, and the result will be a promise of the subclass type. You can customize the type of the return value through the <a href="symbol.species.html"><code>[Symbol.species]</code></a> property.</p>
</div></section><section aria-labelledby="examples"><h2 id="examples">Examples</h2>
</section><section aria-labelledby="using_the_then_method"><h3 id="using_the_then_method">Using the then() method</h3>
<div class="section-content"><div class="code-example"><pre data-language="js">const p1 = new Promise((resolve, reject) =&gt; {
  resolve("Success!");
  // or
  // reject(new Error("Error!"));
});

p1.then(
  (value) =&gt; {
    console.log(value); // Success!
  },
  (reason) =&gt; {
    console.error(reason); // Error!
  },
);
</pre></div></div></section><section aria-labelledby="having_a_non-function_as_either_parameter"><h3 id="having_a_non-function_as_either_parameter">Having a non-function as either parameter</h3>
<div class="section-content"><div class="code-example"><pre data-language="js">Promise.resolve(1).then(2).then(console.log); // 1
Promise.reject(1).then(2, 2).then(console.log, console.log); // 1
</pre></div></div></section><section aria-labelledby="chaining"><h3 id="chaining">Chaining</h3>
<div class="section-content">
<p>The <code>then</code> method returns a new <code>Promise</code>, which allows for method chaining.</p> <p>If the function passed as handler to <code>then</code> returns a <code>Promise</code>, an equivalent <code>Promise</code> will be exposed to the subsequent <code>then</code> in the method chain. The below snippet simulates asynchronous code with the <code>setTimeout</code> function.</p> <div class="code-example"><pre data-language="js">Promise.resolve("foo")
  // 1. Receive "foo", concatenate "bar" to it, and resolve that to the next then
  .then(
    (string) =&gt;
      new Promise((resolve, reject) =&gt; {
        setTimeout(() =&gt; {
          string += "bar";
          resolve(string);
        }, 1);
      }),
  )
  // 2. receive "foobar", register a callback function to work on that string
  // and print it to the console, but not before returning the unworked on
  // string to the next then
  .then((string) =&gt; {
    setTimeout(() =&gt; {
      string += "baz";
      console.log(string); // foobarbaz
    }, 1);
    return string;
  })
  // 3. print helpful messages about how the code in this section will be run
  // before the string is actually processed by the mocked asynchronous code in the
  // previous then block.
  .then((string) =&gt; {
    console.log(
      "Last Then: oops... didn't bother to instantiate and return a promise in the prior then so the sequence may be a bit surprising",
    );

    // Note that `string` will not have the 'baz' bit of it at this point. This
    // is because we mocked that to happen asynchronously with a setTimeout function
    console.log(string); // foobar
  });

// Logs, in order:
// Last Then: oops... didn't bother to instantiate and return a promise in the prior then so the sequence may be a bit surprising
// foobar
// foobarbaz
</pre></div> <p>The value returned from <code>then()</code> is resolved in the same way as <a href="resolve.html"><code>Promise.resolve()</code></a>. This means <a href="../promise.html#thenables">thenable objects</a> are supported, and if the return value is not a promise, it's implicitly wrapped in a <code>Promise</code> and then resolved.</p> <div class="code-example"><pre data-language="js">const p2 = new Promise((resolve, reject) =&gt; {
  resolve(1);
});

p2.then((value) =&gt; {
  console.log(value); // 1
  return value + 1;
}).then((value) =&gt; {
  console.log(value, "- A synchronous value works"); // 2 - A synchronous value works
});

p2.then((value) =&gt; {
  console.log(value); // 1
});
</pre></div> <p>A <code>then</code> call returns a promise that eventually rejects if the function throws an error or returns a rejected Promise.</p> <div class="code-example"><pre data-language="js">Promise.resolve()
  .then(() =&gt; {
    // Makes .then() return a rejected promise
    throw new Error("Oh no!");
  })
  .then(
    () =&gt; {
      console.log("Not called.");
    },
    (error) =&gt; {
      console.error(`onRejected function called: ${error.message}`);
    },
  );
</pre></div> <p>In practice, it is often desirable to <a href="catch.html"><code>catch()</code></a> rejected promises rather than <code>then()</code>'s two-case syntax, as demonstrated below.</p> <div class="code-example"><pre data-language="js">Promise.resolve()
  .then(() =&gt; {
    // Makes .then() return a rejected promise
    throw new Error("Oh no!");
  })
  .catch((error) =&gt; {
    console.error(`onRejected function called: ${error.message}`);
  })
  .then(() =&gt; {
    console.log("I am always called even if the prior then's promise rejects");
  });
</pre></div> <p>In all other cases, the returned promise eventually fulfills. In the following example, the first <code>then()</code> returns <code>42</code> wrapped in a fulfilled Promise, even though the previous Promise in the chain was rejected.</p> <div class="code-example"><pre data-language="js">Promise.reject()
  .then(
    () =&gt; 99,
    () =&gt; 42,
  ) // onRejected returns 42 which is wrapped in a fulfilled Promise
  .then((solution) =&gt; console.log(`Resolved with ${solution}`)); // Fulfilled with 42
</pre></div> <p>If <code>onFulfilled</code> returns a promise, the return value of <code>then</code> will be fulfilled/rejected based on the eventual state of that promise.</p> <div class="code-example"><pre data-language="js">function resolveLater(resolve, reject) {
  setTimeout(() =&gt; {
    resolve(10);
  }, 1000);
}
function rejectLater(resolve, reject) {
  setTimeout(() =&gt; {
    reject(new Error("Error"));
  }, 1000);
}

const p1 = Promise.resolve("foo");
const p2 = p1.then(() =&gt; {
  // Return promise here, that will be resolved to 10 after 1 second
  return new Promise(resolveLater);
});
p2.then(
  (v) =&gt; {
    console.log("resolved", v); // "resolved", 10
  },
  (e) =&gt; {
    // not called
    console.error("rejected", e);
  },
);

const p3 = p1.then(() =&gt; {
  // Return promise here, that will be rejected with 'Error' after 1 second
  return new Promise(rejectLater);
});
p3.then(
  (v) =&gt; {
    // not called
    console.log("resolved", v);
  },
  (e) =&gt; {
    console.error("rejected", e); // "rejected", 'Error'
  },
);
</pre></div> <p>You can use chaining to implement one function with a Promise-based API on top of another such function.</p> <div class="code-example"><pre data-language="js">function fetchCurrentData() {
  // The fetch() API returns a Promise. This function
  // exposes a similar API, except the fulfillment
  // value of this function's Promise has had more
  // work done on it.
  return fetch("current-data.json").then((response) =&gt; {
    if (response.headers.get("content-type") !== "application/json") {
      throw new TypeError();
    }
    const j = response.json();
    // maybe do something with j

    // fulfillment value given to user of
    // fetchCurrentData().then()
    return j;
  });
}
</pre></div>
</div></section><section aria-labelledby="asynchronicity_of_then"><h3 id="asynchronicity_of_then">Asynchronicity of then()</h3>
<div class="section-content">
<p>The following is an example to demonstrate the asynchronicity of the <code>then</code> method.</p> <div class="code-example"><pre data-language="js">// Using a resolved promise 'resolvedProm' for example,
// the function call 'resolvedProm.then(...)' returns a new promise immediately,
// but its handler '(value) =&gt; {...}' will get called asynchronously as demonstrated by the console.logs.
// the new promise is assigned to 'thenProm',
// and thenProm will be resolved with the value returned by handler
const resolvedProm = Promise.resolve(33);
console.log(resolvedProm);

const thenProm = resolvedProm.then((value) =&gt; {
  console.log(
    `this gets called after the end of the main stack. the value received is: ${value}, the value returned is: ${
      value + 1
    }`,
  );
  return value + 1;
});
console.log(thenProm);

// Using setTimeout, we can postpone the execution of a function to the moment the stack is empty
setTimeout(() =&gt; {
  console.log(thenProm);
});

// Logs, in order:
// Promise {[[PromiseStatus]]: "resolved", [[PromiseResult]]: 33}
// Promise {[[PromiseStatus]]: "pending", [[PromiseResult]]: undefined}
// "this gets called after the end of the main stack. the value received is: 33, the value returned is: 34"
// Promise {[[PromiseStatus]]: "resolved", [[PromiseResult]]: 34}
</pre></div>
</div></section><h2 id="specifications">Specifications</h2>
<div class="_table"><table class="standard-table">
<thead><tr><th scope="col">Specification</th></tr></thead>
<tbody><tr><td><a href="https://tc39.es/ecma262/multipage/control-abstraction-objects.html#sec-promise.prototype.then">ECMAScript Language Specification <br><small># sec-promise.prototype.then</small></a></td></tr></tbody>
</table></div>
<h2 id="browser_compatibility">Browser compatibility</h2>
<div class="_table"><table>
<thead>
<tr id="bct-browser-type">
<th></th>
<th colspan="5">Desktop</th>
<th colspan="6">Mobile</th>
<th colspan="2">Server</th>
</tr>
<tr id="bct-browsers">
<th></th>
<th>Chrome</th>
<th>Edge</th>
<th>Firefox</th>
<th>Opera</th>
<th>Safari</th>
<th>Chrome Android</th>
<th>Firefox for Android</th>
<th>Opera Android</th>
<th>Safari on IOS</th>
<th>Samsung Internet</th>
<th>WebView Android</th>
<th>Deno</th>
<th>Node.js</th>
</tr>
</thead>
<tbody><tr>
<th><code>then</code></th>
<td class="bc-supports-yes">32</td>
<td class="bc-supports-yes">12</td>
<td class="bc-supports-yes">29</td>
<td class="bc-supports-yes">19</td>
<td class="bc-supports-yes">8</td>
<td class="bc-supports-yes">32</td>
<td class="bc-supports-yes">29</td>
<td class="bc-supports-yes">19</td>
<td class="bc-supports-yes">8</td>
<td class="bc-supports-yes">2.0</td>
<td class="bc-supports-yes">4.4.3</td>
<td class="bc-supports-yes">1.0</td>
<td class="bc-supports-yes">0.12.0</td>
</tr></tbody>
</table></div>
<section aria-labelledby="see_also"><h2 id="see_also">See also</h2>
<div class="section-content"><ul> <li><a href="../promise.html"><code>Promise</code></a></li> <li><a href="catch.html"><code>Promise.prototype.catch()</code></a></li> </ul></div></section><div class="_attribution">
  <p class="_attribution-p">
    &copy; 2005&ndash;2024 MDN contributors.<br>Licensed under the Creative Commons Attribution-ShareAlike License v2.5 or later.<br>
    <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/then" class="_attribution-link">https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/then</a>
  </p>
</div>
